syntax = "proto3";

package disttx.grpc.proto;

option java_package = "com.tsurugidb.disttx.grpc.proto";
option java_multiple_files = true;

// TP Monitor gRPC service for distributed transaction coordination
service TpMonitorService {
    // Create a new TP monitor instance
    rpc Create(CreateRequest) returns (CreateResponse);

    // Join an existing TP monitor as a participant
    rpc Join(JoinRequest) returns (JoinResponse);

    // Destroy a TP monitor instance (optional, for cleanup)
    rpc Destroy(DestroyRequest) returns (DestroyResponse);

    // Notify barrier arrival (blocks until all participants arrive)
    rpc Barrier(BarrierRequest) returns (BarrierResponse);
}

// Request to create a TP monitor
message CreateRequest {
    uint32 participantCount = 1;
}

// Response for create
message CreateResponse {
    uint64 tpmId = 1;  // TP monitor ID
}

// Request to join a TP monitor
message JoinRequest {
    uint64 tpmId = 1;  // TP monitor ID
    string txId = 2;   // Transaction ID
    uint64 tsId = 3;   // Tsurugi Server ID
}

// Response for join
message JoinResponse {
    bool success = 1;
}

// Request to destroy a TP monitor
message DestroyRequest {
    uint64 tpmId = 1;
}

// Response for destroy
message DestroyResponse {
    bool success = 1;
}

// Request to notify barrier arrival
message BarrierRequest {
    uint64 tpmId = 1;
    string txId = 2;
}

// Response for barrier
message BarrierResponse {
    bool success = 1;
}
